### PR TITLE
dbeaver/dbeaver#23259 SAP IQ procedure definition query has changed and all original indentation is lost

### LINKED ISSUE #23259: SAP IQ procedure definition query has changed and all original indentation is lost
### Description

SAP IQ procedure definition query has changed compared to 24.0.0 and all original indentation is lost

Most likely after https://github.com/dbeaver/dbeaver/commit/2e2e5fafdb46ce10954d70df8a37473d1e0c21bc implementation

### DBeaver Version

CE 24.0.1

### Operating System

_No response_

### Database and driver

SAP IQ, 16.1.050.2532
Driver: jConnect (TM) for JDBC (TM) jConnect (TM) for JDBC(TM)/20.00.000.01(Build 27499)

### Steps to reproduce

Open definition of procedures in 24.0.0 and compare same procedure in 24.0.1

in 24.0.1 original indentation is lost
image

VdvOl1lDI6

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by serjiokov**
Read source procedure when content exists.

### CODE REVIEW

**Comment by serjiokov** [comment]
image

**Comment by LonwoLonwo** [comment]
It is not the correct decision.

File: plugins/org.jkiss.dbeaver.ext.mssql/src/org/jkiss/dbeaver/ext/mssql/model/generic/SQLServerMetaModel.java [hunk 1]
```diff
@@ -172,15 +171,18 @@ public String getProcedureDDL(DBRProgressMonitor monitor, GenericProcedure sourc
...
+                    "SELECT source, proc_defn from SYS.SYSPROCEDURE s\n" +
                         "LEFT JOIN " + DBUtils.getQuotedIdentifier(sourceObject.getCatalog()) + ".dbo.sysobjects so " +
                         "ON so.id = s.object_id\n" +
                         "WHERE s.proc_name=?"
                 )) {
                     dbStat.setString(1, objectName);
                     try (JDBCResultSet dbResult = dbStat.executeQuery()) {
                         if (dbResult.nextRow()) {
-                            return dbResult.getString(1);
+                            if (dbResult.getString(1) != null) {
```
Comment (LonwoLonwo) [inline plugins/org.jkiss.dbeaver.ext.mssql/src/org/jkiss/dbeaver/ext/mssql/model/generic/SQLServerMetaModel.java]:
```
if (dbResult.nextRow()) {
    String originalDefinition = dbResult.getString(1);
    if (CommonUtils.isEmpty(originalDefinition)) {
        // The original routine source code from the ISYSSOURCE system table is empty;
        // we need to use proc_defn, but the database server has parsed and unparsed this definition.
        return dbResult.getString(2);
    }
    return  originalDefinition;
}
```

File: plugins/org.jkiss.dbeaver.ext.mssql/src/org/jkiss/dbeaver/ext/mssql/model/generic/SQLServerMetaModel.java [hunk 2]
```diff
@@ -172,15 +171,18 @@ public String getProcedureDDL(DBRProgressMonitor monitor, GenericProcedure sourc
         if (getServerType() == ServerType.SYBASE) {
```
Comment (LonwoLonwo) [inline plugins/org.jkiss.dbeaver.ext.mssql/src/org/jkiss/dbeaver/ext/mssql/model/generic/SQLServerMetaModel.java:172]: `} else if (getServerType() == ServerType.SYBASE && !isSapIQ(dataSource) && !isASEProduct(dataSource)) {`

Comment (LonwoLonwo) [reply plugins/org.jkiss.dbeaver.ext.mssql/src/org/jkiss/dbeaver/ext/mssql/model/generic/SQLServerMetaModel.java:172]:
```
private boolean isASEProduct(@NotNull GenericDataSource dataSource) {
        String productName = dataSource.getInfo().getDatabaseProductName();
        return CommonUtils.isNotEmpty(productName) && (productName.contains("Adaptive") || productName.contains("ASE"));
}
```

**Comment by serjiokov** [comment]
SAP_1
SAP_2
SAP_3

File: plugins/org.jkiss.dbeaver.ext.mssql/src/org/jkiss/dbeaver/ext/mssql/model/generic/SQLServerMetaModel.java [hunk 3]
```diff
@@ -171,27 +169,52 @@ public String getProcedureDDL(DBRProgressMonitor monitor, GenericProcedure sourc
...
                 throw new DBException(e, dataSource);
             }
         }
         return extractSource(monitor, dataSource, sourceObject, sourceObject.getCatalog(), sourceObject.getSchema().getName(), objectName);
     }
 
+    private boolean isMetaDataViewExists(@NotNull JDBCSession session) {
+        boolean result = false;
+        try (Statement dbStat = session.createStatement()) {
+            try (ResultSet resultSet = dbStat.executeQuery("SELECT 1 FROM SYS.SYSPROCEDURE WHERE 1<>1 LIMIT 1")) {
```
Comment (LonwoLonwo) [inline plugins/org.jkiss.dbeaver.ext.mssql/src/org/jkiss/dbeaver/ext/mssql/model/generic/SQLServerMetaModel.java]: SQL Server doesn't support LIMIT 🙄 
`SELECT TOP 1 1 FROM SYS.SYSPROCEDURE WHERE 1 <> 1`

Comment (serjiokov) [reply plugins/org.jkiss.dbeaver.ext.mssql/src/org/jkiss/dbeaver/ext/mssql/model/generic/SQLServerMetaModel.java]: Ok, I was relying on it
https://infocenter.sybase.com/help/index.jsp?topic=/com.sybase.help.sqlanywhere.12.0.1/dbusage/first-order-formatting.html

Comment (serjiokov) [reply plugins/org.jkiss.dbeaver.ext.mssql/src/org/jkiss/dbeaver/ext/mssql/model/generic/SQLServerMetaModel.java]: done

File: plugins/org.jkiss.dbeaver.ext.mssql/src/org/jkiss/dbeaver/ext/mssql/model/generic/SQLServerMetaModel.java [hunk 4]
```diff
@@ -171,27 +169,53 @@ public String getProcedureDDL(DBRProgressMonitor monitor, GenericProcedure sourc
         }
         if (getServerType() == ServerType.SYBASE) {
             try (JDBCSession session = DBUtils.openMetaSession(monitor, sourceObject, "Read routine definition")) {
-                try (JDBCPreparedStatement dbStat = session.prepareStatement(
-                    "SELECT proc_defn from SYS.SYSPROCEDURE s\n" +
-                        "LEFT JOIN " + DBUtils.getQuotedIdentifier(sourceObject.getCatalog()) + ".dbo.sysobjects so " +
+                if (isMetaDataViewExists(session)) {
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.ext.mssql/src/org/jkiss/dbeaver/ext/mssql/model/generic/SQLServerMetaModel.java]: Let's move it to variable in datasource to callculate it only once per connection

**Review by serge-rider — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.ext.mssql/src/org/jkiss/dbeaver/ext/mssql/model/generic/SQLServerMetaModel.java (hunk 4 above)
Comment (serjiokov) [reply plugins/org.jkiss.dbeaver.ext.mssql/src/org/jkiss/dbeaver/ext/mssql/model/generic/SQLServerMetaModel.java]: Done

**Comment by Matvey16** [comment]
Veified

**Pull request merged**